### DISCUSSION TITLE
[SOLVED] Is the message filter working at all on 3.9.0?

### DISCUSSION THREAD

**Discussion by guihkx** [martinrotter/rssguard#373] [Q&A]
Hey,

So I'm using the following filter to ignore entries with duplicated URLs:

```js
function filterMessage() {
  if (msg.isDuplicateWithAttribute(MessageObject.SameUrl)) {
    return MessageObject.Ignore;
  }
  return MessageObject.Accept;
}
```

However, when I go to "Sample Message" to test this filter and I add an URL which already exists in my database, RSS Guard says the message will be accepted:

![image](https://user-images.githubusercontent.com/626206/110035602-a6715900-7d1a-11eb-95e5-fcf21dcae7d2.png)

Am I doing something wrong here? It used to work just fine on 3.8.4.

Thanks!

**Comment by martinrotter** [comment]
@guihkx Let me check your example.

Also, could you generate debug log while triggering the issue and post here?

**Reply by guihkx** [reply]
Sure, here you go:

https://gist.github.com/guihkx/1214bd30286f82e814c2db907e78bcdb/raw/74da41d7361ca6d910ceb8e461f124813e8d93ea/rssguard-debug.txt

**Reply by martinrotter** [reply]
OK, so.

1. I added feed `http://feeds.feedburner.com/InterestingThingOfTheDay`.
2. I copied your filter contents and assigned filter to the above feed and I copied one of the URLs of messages of the feed to "SAMPLE MESSAGE" for demonstration purposes, in screenshots it is `https://feedpress.me/link/19120/11512369/itotd-on-summer-break`.
3. First thing to note is that message filters dialog when hitting "Test" button behaves differently depending on what node is SELECTED (highlighted, not checked) in feed/category list as the selected node is then used as "source" feed which influences behavior of the filter and can be overriden by option [`MessageObject.AllFeedsSameAccount`](https://github.com/martinrotter/rssguard/blob/master/resources/docs/Message-filters.md#duplicationattributecheck-enum). If you do not select anything, then your SAMPLE MESSAGE "does not have" feed, therefore your "Test" will not work if your filter does **not** use `MessageObject.AllFeedsSameAccount`. Similarly, if you highlight different feed (and still don't use `MessageObject.AllFeedsSameAccount`), your messages will not be properly identified as duplicate, because RSS Guard did not know to which feed they belong - I am fixing the small bug right now. However, if you highlight the exact feed and use one of its messages' URL in SAMPLE MESSAGE, then your filter will work.
4. Note that the above problem does not influence live filter execution on real incoming messages or when you press "Process checked feeds" - in those cases even 3.9.0 should work. Right now I commited the fix to make sure that filters work even for SAMPLE MESSAGE when no feed is highlighted. The above problem should only be part of "Test" use-case.
5. You will probably want to use `MessageObject.AllFeedsSameAccount` flag in your filter to make sure that the duplication check filter takes all feeds into account not just existing message from the same feed as the incoming message?

See below for sample video to demonstrate all use-cases in 3.9.0: https://www.youtube.com/watch?v=pLfYJwHCcvI

See below for sample video with little fix for "Test" feature: https://www.youtube.com/watch?v=do2jVxBN2w4

**Reply by guihkx** [reply]
Thank you so much for the thorough response! It all makes sense now, not sure how I missed the `MessageObject.AllFeedsSameAccount` option, though, so thank you for that too. 😉

**Reply by martinrotter** [reply]
Yes, but in fact there was a bug, so I fixed for version 4.x and also backported to be included in bug-fix release 3.9.1. I will also enhance debug logging for this feature to make sure all parameters are in log too.

**Comment by guihkx** [comment]
Hmm, I think I've got it.

It appears that the "Check All" / "Uncheck All" buttons that are used to select the feeds in which the filter will be applied to, have no effect whatsoever. You have to actually highlight the feed by clicking with your mouse on top of them in order for the filter to work. Not sure if I was able to explain properly, so take a look:

![Peek 2021-03-05 03-05](https://user-images.githubusercontent.com/626206/110074544-279f0f00-7d60-11eb-887b-01ef1e2d4302.gif)

**Reply by martinrotter** [reply]
@guihkx Yes, see my answer above. I made a small fix to make sure that "Existing messages" tab works even when feed is not highlighted. For "SAMPLE MESSAGE" to work, you must highlight the correct feed or make sure to use `MessageObject.AllFeedsSameAccount` message duplication check flag.

**Reply by guihkx** [reply]
It works now. Just beautiful! Yesterday I was getting a couple duplicate articles because I wasn't using `MessageObject.AllFeedsSameAccount`, but now that won't be an issue anymore. I'll mark this discussion as [SOLVED]. Thank you so much for your help!

Ah, here's the updated filter in case someone comes across this:

```js
function filterMessage() {
  if (msg.isDuplicateWithAttribute(MessageObject.SameUrl | MessageObject.AllFeedsSameAccount)) {
    return MessageObject.Ignore;
  }
  return MessageObject.Accept;
}

```

**Reply by martinrotter** [reply]
Yes, exactly. I had even flag for cross-account global checking `MessageObject.AllFeedsAllAccounts` but silently removed it some versions ago as it was just too much. It can be re-added if anyone requested it.

**Reply by guihkx** [reply]
Nah, I actually think the current behavior is fine. I just had to read the documentation properly 😅

**Reply by martinrotter** [reply]
I am thinking it throught and you are actually right. The fix could maybe do more harm than good. I will test with some more specific filters and maybe roll the change back.

**Reply by martinrotter** [reply]
No, nevermind the fix is good. Case closed.

**Reply by martinrotter** [reply]
It even allows per-feed filtering if user does not select any particular feed (when you open the filtering manager, whole "account" is selected and you see ALL messages you have from all feeds combined), now, something like this is possible:

![image](https://user-images.githubusercontent.com/1255302/110076498-dc4b2780-7d84-11eb-81a5-7e4f58dae1c4.png)

**Reply by guihkx** [reply]
That's awesome! I'll definitely play around with it. Thanks <3

**Reply by martinrotter** [reply]
@guihkx Btw just a new feature: I added new "msg" attribute "source" which is floating point number from range 0.0-100.0 which you can use to "sort" your news via message filter. This new "source" attribute also has its own column in message list and fancy "progress bar" icons (higher score, higher progress bar).
![image](https://user-images.githubusercontent.com/1255302/110076647-203e2c80-7d85-11eb-8970-f9b771a5e588.png)
![image](https://user-images.githubusercontent.com/1255302/110076747-4bc11700-7d85-11eb-9c72-0dba5d516c9f.png)

**Reply by guihkx** [reply]
That's something that will come in handy, very nice!
